store/sqlstore/audit_store.go

```diff
@@ -36,19 +35,20 @@ func (s SqlAuditStore) createIndexesIfNotExists() {
 	s.CreateIndexIfNotExists("idx_audits_user_id", "Audits", "UserId")
 }
 
-func (s SqlAuditStore) Save(audit *model.Audit) *model.AppError {
+func (s SqlAuditStore) Save(audit *model.Audit) error {
 	audit.Id = model.NewId()
 	audit.CreateAt = model.GetMillis()
 
 	if err := s.GetMaster().Insert(audit); err != nil {
-		return model.NewAppError("SqlAuditStore.Save", "store.sql_audit.save.saving.app_error", nil, "user_id="+audit.UserId+" action="+audit.Action, http.StatusInternalServerError)
+		return errors.Wrapf(err, "failed to save Audit with userId=%s and action=%s", audit.UserId, audit.Action)
 	}
 	return nil
 }
 
-func (s SqlAuditStore) Get(user_id string, offset int, limit int) (model.Audits, *model.AppError) {
+func (s SqlAuditStore) Get(userId string, offset int, limit int) (model.Audits, error) {
 	if limit > 1000 {
-		return nil, model.NewAppError("SqlAuditStore.Get", "store.sql_audit.get.limit.app_error", nil, "user_id="+user_id, http.StatusBadRequest)
+		// TODO: is it appropiate to set field to limit ?
```

Comment:
Perhaps a new error is in order?

Reply:
Yep, maybe a new error type would be appropiate for those cases

Reply:
@jespino - what do you think?

Reply:
@agnivade @jespino Friendly ping reminder

Reply:
I agree that a different error can be really valuable here. The plain `ErrBadRequest` we can call it `ErrInvalidParams` but follow the `Detail` field approach if we want. But I think I prefer the approach with `extra data` for example, the `input field` and the `input value`, and maybe, a `Detail` too. Provide to the upper level a sensible error that give all the information structured is the best way to relay in the app layer to make the correct decisions (for example, the userId passed, probably is available in the context where the error must be handled). So, I think the invalid param is the most important one (I'm not sure if we need the value at all). the `details` can be built from the ParamName if we want.

Ok, my proposal is, lets go for `ErrInvalidParameter` with `ParamName` field, and the `Error` function is going to build something like `Invalid param "%s" passed to the function`.

What do you think @agnivade and @rvillablanca?

Reply:
Yep, I agree it is valuable to have error with fields giving more context about the problem but there are cases like this when the only one error available is `ErrInvalidInput` but it is coupled to the any entity model (I think), so we can create another one as suggested with other fields and call it `ErrInvalidParams` and that is not the problem. What I think would be valuable is to have a more generic error for cases where `ErrInvalidInput` and `ErrInvalidParams` are not suitable, maybe because the invalid input is not a entity field nor a generic parameter but still being an invalid input. 

This can be extrapolated to any other error, it is more a generic concern and not related only with an invalid input. We need a generic type like `error` but slighly different so we can assert with it and decide if the response should be a bad request, an internal rrror or a not found resource.

Anyway, it is just a suggestion. We can have many kinds of specific errors with different field on them, or have just only a few specific errors plus other ones with `string` detail being different from plain `error` interface.

Reply:
Yes, `ErrInvalidInput` is tied to an invalid value of an entity model. So that definitely does not fit here.

However, before adding a more generic error to capture any invalid input not tied to an entity model, I would really like to see some more practical examples of it in our codebase to make a proper decision.

@rvillablanca - you have converted a lot of these errors and probably have the most experience. Would this be the first time you are seeing an invalid input which is not tied to a domain model? The doubts raised in https://github.com/mattermost/mattermost-server/pull/14836 would again fall into `ErrInvalidInput` category because that's related to the role and scheme respectively.

Do you have any more examples like these?

Reply:
No, I've not seen more other examples for these cases until now. Another case could be `ErrNotFound` where it assumes the search was by id, so when I have a case when some entity was not found but it was searched by multiple criteria then I have to use `fmt.Sprintf`. 

Of course it is not a problem but the point is the more specific the error is, the more difficult is to fill all fields. But I understand your point, maybe we must wait for more cases like these and take the right decision and until now those are the only cases that I've found.

Reply:
Ok, sounds good then. Let's go with `ErrOutOfBounds` with the field name and value for now. If we see more cases, then we can decide on what the right action will be.

Re: `ErrNotFound` great point on that. I think what we can do is make the `id` field to be variadic args of key-value pairs so that multiple fields can be passed. Feel free to send a PR fixing all those `fmt.Sprintf` cases.